### DISCUSSION TITLE
[RFC] Roadmap

### DISCUSSION THREAD

**Discussion by devnix** [dar5hak/vamp#1] [Q&A]
What do you think we should put on the horizon and in which order to archieve a MVP?

Here I have some ideas borrowed from the ones I've shared on Vanat:

- A `CONTRIBUTING` file with some hints
  - Coding style: Elementary's coding style has been suggested (https://docs.elementary.io/develop/writing-apps/code-style)
  - Git commit messages: Do you like this convention? Do you think it's not important at all? https://www.conventionalcommits.org/en/v1.0.0/
  - When closing a PR, should we squash it first?
- JSON schema for packages, and repositories 
  - Seems like we have decided we like this specification, so we could write the schema for that one: https://github.com/valum-framework/drakkar
- Will we do TDD? (I have to learn a little more by example personally but I would totally advocate for it)
- Where/how are packages going to be published? We could abstract the repository concept for the package manager, so it ignores when it's VCS, central repository, etc (see https://getcomposer.org/doc/05-repositories.md)
- Dependency tree resolution, I guess it should be able to download dependencies recursively, detect conflicts, and try to get the latest compatible version of each possible package (or the lowest one if specified)

**Comment by dar5hak** [comment]
@devnix for starting the conversation!

- **Contributing guidelines:** I agree with all three of your points. But for commit message convention, we'd recommend it without being too pedantic about it.
- **JSON schema:** Yes, we'll try to follow Drakkar in general and deviate when needed.
- **TDD:** I don't have an answer yet, but most probably, we'll try and have unit tests without necessarily them being written before dev as TDD suggests. I don't have much experience in that area either.
- Packages will not be published centrally anywhere, at least at first. We're planning to do it like Golang, where devs point to git repos. We'll see how that goes.
- **Dependency resolution:** yes.

Another item on the roadmap is integration with Flatpak and Meson.

For Flatpak, when a user installs a dep, we plan to automatically add the necessary module info (along with subdeps) to the manifest.

Meson integration is TBD.

Others, please share your thoughts, and correct me if I'm missing anything.

**Reply by dar5hak** [reply]
Oh, and we also plan to have an init command, which generates a JSON for you, based on some questions.

**Reply by devnix** [reply]
The whole building tools integration is another big subject here, I think that's 100% out of the scope of the package manager, but facilitating to those tools things like the list of files to compile, vapis, flags, etc looks like one of the nicest features

**Reply by nyx-lyb3ra** [reply]
I consider the integration with Flatpak to be the most important point and it should be blocking for our first release, since our main users will be the people of elementary

It should be noted that within a Flatpak compilation environment there is no internet access, so Vamp will not be able to download the packages on its own
